### PR TITLE
Downloaded data integrity

### PR DESCRIPTION
**Pull request by lcaparroz**
## Why

In order to assure downloaded data integrity between database, server disk and repository metadata, we need to start tracking all files in the database, even if they were just deduplicated.

## What

- [x] the uniqueness constraint on the index `checksum + checksum_type` on the `downloaded_files` table, in order to allow tracking deduplicated files.
- [x] Start to tracking files on deduplication (copy/hardlink file and then add it to the database).
- [x] Add files to the database if they've been deduplicated previously but are untracked.
- [x] ~Enforce uniqueness to the `local_path` field on the `downloaded_files` table; there shouldn't be two entries with the same local path.~ Add *on the fly* duplicates check to assure there is only one entry for a given local path (P.S.: enforcing uniqueness on the database could cause disruption to the customer, since we can't assert there'll be no duplicates prior to this change). 

## Considerations

We decided to maintain the data consistency during the mirroring process instead of providing separated commands or flags.

### Assumptions

* The repository metadata is the authoritative source of information (e.g.: if file on disk and on database matches, but the checksum is different from the repository metadata, file/database is corrupted);
* RMT should always track files, unless `airgap_mode` is enabled.

### Decision logic for file tracking

Here is a summary of how RMT decides whether it should download/deduplicate/track files or not.

#### Conditions

1. File with given path exists on disk
2. File on disk matches metadata checksum
3. Entry on DB w/ metadata path exists on database
4. Entry on DB w/ metadata checksum exists on database

| 1 | 2 | 3 | 4 | Status | Action |
|:----------------:|:----------------:|:----------------:|:----------------:|:----------------------------|:---------------------------------------|
|:x:               |:x:               |:x:               |:x:               | New file                    | Download -> Track                      |
|:x:               |:x:               |:x:               |:heavy_check_mark:| Same file on another path   | Deduplicate -> Track                   |
|:x:               |:x:               |:heavy_check_mark:|:x:               | Missing on disk             | Untrack -> Download -> Track           |
|:x:               |:x:               |:heavy_check_mark:|:heavy_check_mark:| Missing on disk             | Untrack -> Download -> Track           |
|:x:               |:heavy_check_mark:|:x:               |:x:               | (Impossible)                | ---                                    |
|:x:               |:heavy_check_mark:|:x:               |:heavy_check_mark:| (Impossible)                | ---                                    |
|:x:               |:heavy_check_mark:|:heavy_check_mark:|:x:               | (Impossible)                | ---                                    |
|:x:               |:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:| (Impossible)                | ---                                    |
|:heavy_check_mark:|:x:               |:x:               |:x:               | Corrupted untracked file    | Remove -> Download -> Track            |
|:heavy_check_mark:|:x:               |:x:               |:heavy_check_mark:| Corrupted untracked file    | Remove -> Deduplicate -> Track         |
|:heavy_check_mark:|:x:               |:heavy_check_mark:|:x:               | Corrupted file/DB           | Remove -> Untrack -> Download -> Track |
|:heavy_check_mark:|:x:               |:heavy_check_mark:|:heavy_check_mark:| Corrupted file              | Remove -> Untrack -> Download -> Track |
|:heavy_check_mark:|:heavy_check_mark:|:x:               |:x:               | Untracked file              | Track                                  |
|:heavy_check_mark:|:heavy_check_mark:|:x:               |:heavy_check_mark:| Untracked deduplicated file | Track                                  |
|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:x:               | Corrupted DB                | Untrack -> Track                       |
|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:| Consistent data             | Ignore (clean data)                    |

### CODE REVIEW

**Comment by suse-tests-pass** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Unless this is a trivial change, please include a CHANGELOG entry.<br />
Run <code>osc vc</code> in the <code>package</code> directory to add one.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/122432/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 4ebb72845daadddbd43d41c2ba2f8491c9d33c0f into ea2b2116d04bd33b3855ab1d5dbac7410d708810)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/124726/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 253370439bc04f1d635129208b5c39dfc8ec103c into ea2b2116d04bd33b3855ab1d5dbac7410d708810)

File: lib/rmt/downloader.rb [hunk 1]
```diff
@@ -181,17 +172,38 @@ def finalize_download(request, local_file, checksum_type = nil, checksum_value =
...
+                                size: File.size(local_file))
+    end
 
     @logger.info("↓ #{File.basename(local_file)}")
   rescue StandardError => e
     request.download_path.unlink
     raise e
   end
 
+  def verify_checksum!(checksum_type, checksum_value, download_path)
```
Comment (jesusbv) [inline lib/rmt/downloader.rb:193]: Shouldn't this method remain in `RMT::ChecksumVerifier` ?

It looks like the reponsability to verify a checksum should be in `ChecksumVerifier`.

File: lib/tasks/encrypted_key.rake [hunk 2]
```diff
@@ -2,16 +2,16 @@ namespace :rmt do
   namespace :secrets do
     desc 'Create encryption key for Rails secrets'
     task create_encryption_key: :environment do
-      require "rails/generators/rails/encryption_key_file/encryption_key_file_generator"
+      require 'rails/generators/rails/encryption_key_file/encryption_key_file_generator'
```
Comment (jesusbv) [inline lib/tasks/encrypted_key.rake:5]: As good as it is to remove rubocop offenses and have the code in a good state, this should be in a different PR.

File: lib/rmt/downloader.rb (hunk 1 above)
Comment (lcaparroz) [reply lib/rmt/downloader.rb:193]: This method only wraps `RMT::ChecksumVerifier.match_checksum?` to have extra behavior:

1. If there's no checksum information, it just ignores the verification;
1. If the file's checksum doesn't match the expected digest, it raises an exception.

In this PR I refactored the class `RMT::ChecksumVerifier` to change its behavior: previously it had the method `.verify_checksum` which would raise an exception every time a mismatch happened. However, I needed to reuse the verification logic in other places where I didn't need an exception, so I refactored the method to have a boolean return type and decided what to do with it on each class that depends on checksum verifications.

`RMT::Downloader` has its own context: sometimes we download a file (mainly `.xml` repository metadata) and we don't need to verify checksums. If you take a look at the other methods, you'll see some of them receive checksums as optional parameters, with default to `nil`.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/126339/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged b3a42ff1d3c1a194c9f4c0d35eaa5225e154f543 into ea2b2116d04bd33b3855ab1d5dbac7410d708810)

File: app/models/downloaded_file.rb [hunk 3]
```diff
@@ -1,18 +1,47 @@
...
-    DownloadedFile.find_by({ checksum_type: checksum_type, checksum: checksum })
-  end
+    def get_local_path_by_checksum(checksum_type, checksum)
+      DownloadedFile.find_by({ checksum_type: checksum_type, checksum: checksum })
+    end
+
+    def valid_local_file?(checksum_type, checksum_value, path)
+      return false unless File.exist?(path)
+
+      valid_checksum = RMT::ChecksumVerifier.match_checksum?(checksum_type, checksum_value, path)
```
[the remark this answers is not in the document]
Comment (lcaparroz) [reply app/models/downloaded_file.rb]: I'm going to implement this in a new PR.

File: lib/rmt/mirror.rb [hunk 4]
```diff
@@ -178,28 +180,33 @@ def replace_directory(source_dir, destination_dir)
...
-        parsed_file
-      end
+
+      valid_local_file?(parsed_file, local_file) || deduplicate(parsed_file, local_file)
     end
-    files.compact
+  end
+
+  def valid_local_file?(parsed_file, destination)
+    ::DownloadedFile.valid_local_file?(parsed_file.checksum_type, parsed_file.checksum, destination)
```
[the remark this answers is not in the document]
Comment (lcaparroz) [reply lib/rmt/mirror.rb]: Removed :heavy_check_mark:

File: lib/rmt/mirror.rb [hunk 5]
```diff
@@ -178,28 +180,33 @@ def replace_directory(source_dir, destination_dir)
...
     @logger.info("→ #{File.basename(destination)}")
     true
   rescue ::RMT::Deduplicator::MismatchException => e
     @logger.debug(_('× File does not exist or has wrong filesize, deduplication ignored %{error}.') % { error: e.message })
     false
   end
 
-  def parsed_files_after_dedup(root_path, referenced_files)
-    files = referenced_files.map do |parsed_file|
+  def filter_downloadable_files(root_path, referenced_files)
```
[the remark this answers is not in the document]
Comment (lcaparroz) [reply lib/rmt/mirror.rb]: Implemented on pr #594.

File: lib/tasks/encrypted_key.rake (hunk 2 above)
Comment (lcaparroz) [reply lib/tasks/encrypted_key.rake:5]: Sure! Usually our pipeline don't pass when there are Rubocop offenses, I really don't know how this one got into our main branch.

File: lib/rmt/downloader.rb (hunk 1 above)
Comment (jesusbv) [reply lib/rmt/downloader.rb:193]: > This method only wraps `RMT::ChecksumVerifier.match_checksum?` to have extra behavior:
> 
>     1. If there's no checksum information, it just ignores the verification;
> 
>     2. If the file's checksum doesn't match the expected digest, it raises an exception.
> 
> 
> In this PR I refactored the class `RMT::ChecksumVerifier` to change its behavior: previously it had the method `.verify_checksum` which would raise an exception every time a mismatch happened. However, I needed to reuse the verification logic in other places where I didn't need an exception, so I refactored the method to have a boolean return type and decided what to do with it on each class that depends on checksum verifications.

That makes sense to me and I think is a good change and this explanation helps.
> 
> `RMT::Downloader` has its own context: sometimes we download a file (mainly `.xml` repository metadata) and we don't need to verify checksums. If you take a look at the other methods, you'll see some of them receive checksums as optional parameters, with default to `nil`.

Please, correct me if I'm wrong, the method is used once [here](https://github.com/SUSE/rmt/pull/585/files#diff-b5294cd24ec48cddb9d6a240471ec8c4R175) to raise an exception (or not), so maybe a rename to `handle_checksum_verification` (or something more accurate of what the method does) or maybe avoid the method.

File: app/models/downloaded_file.rb [hunk 6]
```diff
@@ -1,18 +1,17 @@
...
-    nil
+        record.save if record.changed?
+      end
+    end
   end
 
-  def self.get_local_path_by_checksum(checksum_type, checksum)
-    DownloadedFile.find_by({ checksum_type: checksum_type, checksum: checksum })
+  def size
+    file_size
```
Comment (gionaufal) [inline app/models/downloaded_file.rb:15]: wouldn't be better to use a proper alias here? or even more, why do we need this method/alias?

Comment (lcaparroz) [reply app/models/downloaded_file.rb:15]: I tried using the Ruby `alias` but if failed to recognize the model attributes from `ActiveRecord`. There is [`alias_attribute`](https://apidock.com/rails/Module/alias_attribute) (which I just found out), and it works fine.

The alias makes the model instance interchangeable with `RMT::Mirror::FileReference` instances in `RMT::FileValidator`; I could have named the attribute on the later `file_size` too, but it seemed redundant. Even in the model itself it's a bit redundant, since the table is called `downloaded_files` and other file attributes are not prefixed by `file_`. If it's confusing to have the alias, I can remove it and implement a similar interface in `RMT::Mirror::FileReference` instead.

What do you think?

Comment (gionaufal) [reply app/models/downloaded_file.rb:15]: Cool, thanks for explaining! Makes sense for me.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/126591/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged b3a42ff1d3c1a194c9f4c0d35eaa5225e154f543 into 0354adf79aa7d6843f76cf4f1ed432a0f329a603)

File: lib/tasks/encrypted_key.rake (hunk 2 above)
Comment (lcaparroz) [reply lib/tasks/encrypted_key.rake]: Another PR merged today implemented the same changes.

File: lib/rmt/downloader.rb (hunk 1 above)
Comment (lcaparroz) [reply lib/rmt/downloader.rb]: I renamed the method to `handle_checksum_verification` in the last commit.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/126606/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged c2afd7edfaadca337c1285c7c22f635172053e42 into 0354adf79aa7d6843f76cf4f1ed432a0f329a603)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/126622/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 43d4ca11247a43dc43993a9b5940a44ae9935212 into 0354adf79aa7d6843f76cf4f1ed432a0f329a603)
